### PR TITLE
feat: slash commands and interactions

### LINKED ISSUE #247: Support for Slash Commands and Interactions
Discord will be adding slash commands soon. Documentation PR is [here](https://github.com/discord/discord-api-docs/pull/2295) already.

Discordeno should implement that when its out.

To-Do:
- [x] Events
- [x] Controllers
- [x] Handlers
- [x] Structures
- [x] Types

### PR DESCRIPTION
**Pull request by Skillz4Killz**
**Please describe the changes this PR makes and why it should be merged:**
- [x] Added slash commands 
- [x] Added interactions

Reference: https://github.com/discord/discord-api-docs/pull/2295

Closes #247 

**Status**

- [x] Code changes have been tested against the Discord API, or there are no code changes

**Semantic versioning classification:**

- [x] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

**Comment by ayntee** [comment]
Rename `registerSlashCommand` to `createSlashCommand` to increase the predictability?

File: src/module/interactions.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+// TODO: https://github.com/discord/discord-api-docs/pull/2295/files#diff-07978ef9d619062fb1afbb70ea065124531246743444f4354cc9e6c8879f8780R256
+
+// 1. Your endpoint must be prepared to ACK a `PING` message
+// 2. Your endpoint must be set up to properly handle signature headers--more on that in [Security and Authorization]
+
+const serverOptions = {
+  hostname: "localhost",
+  // Made of our favorite numbers for the default port
+  port: 0704,
```
Comment (itohatweb) [inline src/module/interactions.ts:11]: I do not think ports can start with 0

**Review by itohatweb — CHANGES_REQUESTED** [review]

File: src/module/interactions.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import { serveTLS } from "https://deno.land/std@0.80.0/http/server.ts";
+
+// TODO: https://github.com/discord/discord-api-docs/pull/2295/files#diff-07978ef9d619062fb1afbb70ea065124531246743444f4354cc9e6c8879f8780R256
+
+// 1. Your endpoint must be prepared to ACK a `PING` message
+// 2. Your endpoint must be set up to properly handle signature headers--more on that in [Security and Authorization]
+
+const serverOptions = {
+  hostname: "localhost",
```
Comment (itohatweb) [inline src/module/interactions.ts:9]: Consider changing to 0.0.0.0 in order to listen on every network interface.

File: src/module/interactions.ts (hunk 1 above)
Comment (Skillz4Killz) [reply src/module/interactions.ts:11]: Its just a placeholder waiting for ayntee favorite number

File: src/module/interactions.ts (hunk 2 above)
Comment (itohatweb) [reply src/module/interactions.ts:9]: Just tested, server will only listen to requests coming from own machine.

File: src/module/interactions.ts (hunk 1 above)
Comment (ayntee) [reply src/module/interactions.ts:11]: `1729` :joy:

**Comment by ayntee** [comment]
It would be great if we can create a guide page on slash commands for beginners!

**Comment by itohatweb** [comment]
First Slash Commands should be added to the boilerplate. Since docs are about the boilerplate.

**Comment by ayntee** [comment]
> First Slash Commands should be added to the boilerplate. Since docs are about the boilerplate.

Like we previously discussed, the library should have its own standalone guide. But for now, we can add "slash commands & interactions" in the boilerplate guide, formally referred as "the guide" .

**Comment by Skillz4Killz** [comment]
TODO:

1. Create a boilerplate repo for just the https side version of slash commands
2. Create a ratelimit system for the https side as well
3. Create handlers for the endpoints for the https side.

File: src/interactions/mod.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+async function handlePayload(payload: Interaction) {
+  switch (payload.type) {
+    case InteractionType.PING:
+      return { status: 200, body: { type: InteractionResponseType.PONG } };
+    default: // APPLICATION_COMMAND
+      return controllers.handleApplicationCommand(payload);
+  }
+}
+
+/** The function that handles your commands. */
```
Comment (ayntee) [inline src/interactions/mod.ts:81]: This needs a better description.

File: src/interactions/mod.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      type: InteractionResponseType.CHANNEL_MESSAGE_WITH_SOURCE,
+      data: {
+        content:
+          "Whoopsies! Seems the handling for this command is missing. Please contact my developers!",
+      },
+    },
+  };
+}
+
+/** Internal function to verify security */
```
Comment (ayntee) [inline src/interactions/mod.ts:108]: This needs a better description.

File: src/interactions/interactions.ts [hunk 5]
```diff
@@ -7,7 +7,7 @@ import {
 } from "./types/mod.ts";
 
 /** This variable is a holder for the public key */
-const serverOptions = {
+export let serverOptions = {
```
Comment (Skillz4Killz) [inline src/interactions/interactions.ts]: why was this changed?

**Comment by Skillz4Killz** [comment]
This should be ready for final reviews and cleanup.
